src/cpu/testers/traffic_gen/gups_gen.cc

```diff
@@ -0,0 +1,414 @@
+    system(params.system),
+    requestorId(system->getRequestorId(this)),
+    port(name() + ".port", this),
+    blockSize(params.block_size),
+    startAddr(params.start_addr),
+    memSize(params.mem_size),
+    reqQueueSize(params.request_queue_size),
+    doGenerate(true),
+    doneReading(false),
+    onTheFlyRequests(0),
+    elementSize(8),
+    stats(this)
+{}
+
+Port&
+GUPSGen::getPort(const std::string &if_name, PortID idx)
+{
+    if (if_name != "port"){
+        return ClockedObject::getPort(if_name, idx);
+    }
+    else{
+        return port;
+    }
+}
+
+void
+GUPSGen::init()
+{
+
+    tableSize = memSize / sizeof(uint64_t);
+    numUpdates = 4 * tableSize;
+
+    for (int i = 0; i < 128; i++)
+    {
+        randomSeeds[i] = randomNumberGenerator((numUpdates / 128) * i);
+    }
+
+}
+
+void
+GUPSGen::startup()
+{
+    uint64_t stride_size = blockSize / sizeof(uint64_t);
+
+    for (uint64_t start_index = 0; start_index < tableSize;
+                                start_index += stride_size)
+    {
+        uint8_t write_data[blockSize];
+        for (uint64_t offset = 0; offset < stride_size; offset++)
+        {
+            uint64_t value = start_index + offset;
+            std::memcpy(write_data + offset * sizeof(uint64_t),
+                        &value, sizeof(uint64_t));
+        }
+        Addr addr = indexToAddr(start_index);
+        PacketPtr pkt = getWritePacket(addr, blockSize, write_data);
+        port.sendFunctionalPacket(pkt);
+        delete pkt;
+    }
+
+    schedule(nextSendEvent, nextCycle());
+
+}
+
+uint64_t
+GUPSGen::randomNumberGenerator(int64_t n)
+{
+    int i, j;
+    uint64_t m2[64];
+    uint64_t temp, ran;
+
+    while (n < 0) n += PERIOD;
+    while (n > PERIOD) n -= PERIOD;
+    if (n == 0) return 0x1;
+
+    temp = 0x1;
+    for (i = 0; i < 64; i++)
+    {
+        m2[i] = temp;
+        temp = (temp << 1) ^ ((int64_t) temp < 0 ? POLY : 0);
+        temp = (temp << 1) ^ ((int64_t) temp < 0 ? POLY : 0);
+    }
+
+    for (i = 62; i >= 0; i--)
+    {
+        if ((n >> i) & 1)
+            break;
+    }
+
+    ran = 0x2;
+    while (i > 0)
+    {
+        temp = 0;
+        for (j = 0; j < 64; j++)
+            if ((ran >> j) & 1)
+            temp ^= m2[j];
+        ran = temp;
+        i -= 1;
+        if ((n >> i) & 1)
+            ran = (ran << 1) ^ ((int64_t) ran < 0 ? POLY : 0);
+    }
+
+    return ran;
+}
+
+Addr
+GUPSGen::indexToAddr (uint64_t index)
+{
+    Addr ret = index * elementSize + startAddr;
+
+    return ret;
+}
+
+PacketPtr
+GUPSGen::getReadPacket(Addr addr, unsigned int size)
+{
+    RequestPtr req = std::make_shared<Request>(addr, size,
+                                            0, requestorId);
+    // Dummy PC to have PC-based prefetchers latch on; get entropy into higher
+    // bits
+    req->setPC(((Addr)requestorId) << 2);
+
+    // Embed it in a packet
+    PacketPtr pkt = new Packet(req, MemCmd::ReadReq);
+
+    uint8_t* pkt_data = new uint8_t[req->getSize()];
```

Comment:
I don't think you need to do this. I think you can ask the pkt to allocate the memory itself (then you don't have to remember to delete it)

Reply:
ping